src/Graphs/DirectedGraphBuilder.cs

```diff
@@ -0,0 +1,210 @@
+        private readonly Dictionary<T, HashSet<T>> directPredecessors = new Dictionary<T, HashSet<T>>();
+
+        public static DirectedGraphBuilder<T> NewBuilder()
+        {
+            return new DirectedGraphBuilder<T>();
+        }
+
+        public static DirectedGraphBuilder<T> FromExistingGraph(IDirectedGraph<T> graph)
+        {
+            var builder = new DirectedGraphBuilder<T>();
+
+            foreach (var element in graph.Elements)
+            {
+                builder.elements.Add(element);
+
+                var successors = new HashSet<T>(graph.GetDirectSuccessorsOf(element));
+                var predecessors = new HashSet<T>(graph.GetDirectPredecessorsOf(element));
+
+                builder.directSuccessors.Add(element, successors);
+                builder.directPredecessors.Add(element, predecessors);
+                if (predecessors.Count == 0)
+                {
+                    builder.sources.Add(element);
+                }
+            }
+
+            return builder;
+        }
+
+        public static IDirectedAcyclicGraph<T> EmptyGraph()
+        {
+            return new DirectedGraphBuilder<T>().CreateAcyclicGraphUnsafe();
+        }
+
+        private DirectedGraphBuilder() { }
+
+        public DirectedGraphBuilder<T> AddElement(T element)
+        {
+            if (element == null) throw new ArgumentNullException(nameof(element));
+
+            if (!elements.Add(element))
+            {
+                return this;
+            }
+            sources.Add(element);
+            directSuccessors.Add(element, new HashSet<T>());
+            directPredecessors.Add(element, new HashSet<T>());
+
+            return this;
+        }
+
+        public DirectedGraphBuilder<T> AddArrow(T from, T to)
+        {
+            if (from == null) throw new ArgumentNullException(nameof(from));
+            if (to == null) throw new ArgumentNullException(nameof(to));
+            if (!elements.Contains(from)) throw new ArgumentException("Element not in graph.", nameof(from));
+            if (!elements.Contains(to)) throw new ArgumentException("Element not in graph.", nameof(to));
+            if (from.Equals(to))
+            {
+                throw new InvalidOperationException("Cannot make self-edges in directed acyclic graph.");
+            }
+
+            directSuccessors[from].Add(to);
+            directPredecessors[to].Add(from);
+
+            if (sources.Contains(to))
+            {
+                sources.Remove(to);
+            }
+
+            return this;
+        }
+
+        public DirectedGraphBuilder<T> RemoveElement(T element)
+        {
+            if (element == null) throw new ArgumentNullException(nameof(element));
+
+            if (!elements.Remove(element))
+            {
+                return this;
+            }
+            sources.Remove(element);
+
+            foreach (var to in directSuccessors[element])
```

Comment:
I could probably understand this loop with more thinking or more sleep prior to reviewing, but I suggest it might want to have a name.